src/seedlink/apps/3rd-party/slarchive/doc/slarchive.1
.TH SLARCHIVE 1 2005/08/25
.SH NAME
slarchive \- SeedLink client for data stream archiving

.SH SYNOPSIS
.nf
slarchive [options] [host][:][port]

.fi
.SH DESCRIPTION
\fBslarchive\fP connects to a \fISeedLink\fR server, requests data
streams and writes received packets into directory/file structures.
The precise layout of the directories and files is defined in a
\fIformat\fP string.  Some preset file layouts are included: the
SeisComP Data Structure (SDS), Buffer of Uniform Data structure (BUD)
and the old SeisComP/datalog structure (DLOG) for backwards
compatibility.  To write more than one archive simply specify multiple
\fIformat\fP definitions (or presets).

.SH OPTIONS

.IP "-V         "
Print program version and exit.

.IP "-h         "
Print program usage and exit.

.IP "-v         "
Be more verbose.  This flag can be used multiple times ("-v -v" or
"-vv") for more verbosity.  One flag: report basic handshaking (link
configuration) details and briefly report each packet received.  Two
flags: report the details of the handshaking, each packet received and
detailed connection diagnostics.

.IP "-p             "
Print details of received Mini-SEED data records. This flag can be
used multiple times ("-p -p" or "-pp") for more detail.  One flag: a
single summary line for each data packet received.  Two flags: details
of the Mini-SEED data records received, including information from
fixed header and 100/1000/1001 blockettes.

.IP "-nd \fIdelay\fR"
The network reconnect delay (in seconds) for the connection to
the SeedLink server.  If the connection breaks for any reason
this will govern how soon a reconnection should be attempted.
The default value is 30 seconds.

.IP "-nt \fItimeout\fR"
The network timeout (in seconds) for the connection to the SeedLink
server.  If no data [or keep alive packets?] are received in this 
time the connection is closed and re-established (after the 
reconnect delay has expired).  The default value is 600 seconds.
A value of 0 disables the timeout.

.IP "-k \fIkeepalive\fR  (requires SeedLink >= 3)"
Keepalive packet interval (in seconds) at which keepalive (heartbeat)
packets are sent to the server.  Keepalive packets are only sent if
nothing is received within the interval.

.IP "-x \fIstatefile\fR[:\fIinterval\fR]"
During client shutdown the last received sequence numbers and time
stamps (start times) for each data stream will be saved in this file.
If this file exists upon startup the information will be used to
resume the data streams from the point at which they were stopped.  In
this way the client can be stopped and started without data loss,
assuming the data are still available on the server.  If
\fIinterval\fR is specified the state will be saved every
\fIinterval\fR packets that are received.  Otherwise the state
will be saved only on normal program termination.

.IP "-i \fItimeout\fR"
Timeout for closing idle data stream files in seconds.  The idle time
of data streams is only checked when a packet has arrived so if no
packets are arriving no idle stream files will be closed.  There
should be no reason to change this parameter except for unusual cases
where the process is running against an open file number limit.
Default is 300 seconds.

.IP "-d"
Configure the connection in "dial-up" mode.  The remote server will
close the connection when it has sent all of the data in it's buffers
for the selected data streams.  This is opposed to the normal behavior
of waiting indefinately for data.

.IP "-Fi[:overlap]  "
Future check initially.  Check the last Mini-SEED data record in an
existing archive file and do not write new data to that file if it is
older within a certain overlap.  The default overlap limit is 2
seconds; the overlap can be specified by appending a colon and the
desired overlap limit in seconds to the option.  If the overlap is
exceeded an error message will be logged once for each time the file
is opened.  This option only makes sense for archive formats where
each unique data stream is written to a unique file (e.g. SDS format).
If a data stream is closed due to timeout (see option -i) the initial
future check will be preformed again if the file is re-opened.

.IP "-Fc[:overlap]  "
Future check continuously.  Only archive Mini-SEED data records if the
first sample of the record is newer, within a certain overlap, than
the last sample of the previous record for a given archive file.  The
default overlap limit is 2 seconds; the overlap can be specified by
appending a colon and the desired overlap limit in seconds to the
option.  If the overlap is exceeded an error message will be logged
once until either a non-overlapping packet is received or a new
archive file is used.  This option only makes sense for archive
formats where each unique data stream is written to a unique file
(e.g. SDS format).

.IP "-A \fIformat\fR"
If specified, all packets (Mini-SEED records) received will be
appended to a directory/file structure defined by \fBformat\fP.  All
directories implied in the \fBformat\fP string will be created if
necessary.  The option may be used multiple times to write received
packets to multiple archives.  See the section \fIArchiving data\fR.

.IP "-SDS \fISDSdir\fR"
If specified, all packets (Mini-SEED records) received will be saved
into a Simple Data Structure (SDS) dir/file structure starting at the
specified directory.  This directory and all subdirectories will be
created if necessary.  This option is essentially a preset version of
the '-A' option.  The SDS dir/file structure is:
.nf
<SDSdir>/<YEAR>/<NET>/<STA>/<CHAN.TYPE>/NET.STA.LOC.CHAN.TYPE.YEAR.DAY
.fi

.IP "-BUD \fIBUDdir\fR"
If specified, all waveform data packets (Mini-SEED data records)
received will be saved into a Buffer of Uniform Data (BUD) dir/file
structure starting at the specified directory.  This directory and all
subdirectories will be created if necessary.  This option is
essentially a preset version of the '-A' option.  The BUD dir/file
structure is:
.nf
<BUDdir>/<NET>/<STA>/STA.NET.LOC.CHAN.YEAR.DAY
.fi

.IP "-DLOG \fIDLOGdir\fR"
If specified, all packets (Mini-SEED data records) received will be
saved into an old style SeisComP/datalog dir/file structure starting
at the specified directory.  This directory and all subdirectories
will be created if necessary.  This option is essentially a preset
version of the '-A' option.  The DLOG dir/file structure is:
.nf
<DLOGdir>/<STA>/[LOC.]<CHAN>.<TYPE>/STA.NET.CHAN.TYPE.YEAR.DAY.HHMM
.fi

.IP "-l \fIstreamfile\fR"
A list of streams will be read from the given file.  This option
implies multi-station mode.  The format of the stream list file is
given below in the section \fIStream list file\fR.

.IP "-s \fIselectors\fR"
This defines default selectors.  If no multi-station data streams are
configured these selectors will be used for uni-station mode.
Otherwise these selectors will be used when no selectors are specified
for a given stream using the '-S' or '-l' options.

.IP "-S \fIstream[:selectors],...\fR  (requires SeedLink >= 2.5)"
The connection will be configured in multi-station mode with optional
SeedLink selectors for each station, see examples below.  \fIstream\fR
should be provided in NET_STA format.  If no selectors are provided
for a given stream, the default selectors, if defined, will be used.

.IP "-tw \fIstart:[end]\fR  (requires SeedLink >= 3)"
Specifies a time window applied, by the server, to data streams.  The
format for both times is year,month,day,hour,min,sec; for example:
"2002,08,05,14,00:2002,08,05,14,15,00".  The end time is optional but
the colon must be present.  If no end time is specified the server
will send data indefinately.  This option will override any saved
state information.

Warning: time windowing might be disabled on the remote server.

.IP "\fI[host][:][port]\fR"
A required argument, specifies the address of the SeedLink server in
host:port format.  Either the host, port or both can be omitted.  If
host is omitted then localhost is assumed, i.e.  ':18000'
implies 'localhost:18000'.  If the port is omitted then 18000 is
assumed, i.e.  'localhost' implies 'localhost:18000'.  If only ':' is
specified 'localhost:18000' is assumed.

.SH "SeedLink SELECTORS"
\fINotes regarding selectors from a SeedLink configuration file\fR

The "selectors" parameter is used to request specific packets,
in effect limiting the default action of sending all data.
A packet is sent to the client if it matches any positive selector
(without leading "!") and doesn't match any negative selectors
(with "!").  The general format of selectors is LLSSS.T, where LL is
location, SSS is channel, and T is type (one of DECOTL for Data,
Event, Calibration, Blockette, Timing, and Log records).  "LL",
".T", and "LLSSS." can be omitted, meaning "any".  It is
also possible to use "?" in place of L and S.

.nf

Some examples:
BH?            - BHZ, BHN, BHE (all record types)
00BH?.D        - BHZ, BHN, BHE with location code '00' (data records)
BH? !E         - BHZ, BHN, BHE (excluding detection records)
BH? E          - BHZ, BHN, BHE & detection records of all channels
!LCQ !LEP      - exclude LCQ and LEP channels
!L !T          - exclude log and timing records
.fi

.SH "Archiving data"
Using the '-A \fBformat\fP' option received data can be saved in a
custom directory and file structure.  The archive \fBformat\fP
argument is expanded for each packet processed using the following
flags:

.nf
  \fBn\fP : network code, white space removed
  \fBs\fP : station code, white space removed
  \fBl\fP : location code, white space removed
  \fBc\fP : channel code, white space removed
  \fBY\fP : year, 4 digits
  \fBy\fP : year, 2 digits zero padded
  \fBj\fP : day of year, 3 digits zero padded
  \fBH\fP : hour, 2 digits zero padded
  \fBM\fP : minute, 2 digits zero padded
  \fBS\fP : second, 2 digits zero padded
  \fBF\fP : fractional seconds, 4 digits zero padded
  \fB%\fP : the percent (%) character
  \fB#\fP : the number (#) character
  \fBt\fP : single character type code:
         D - waveform data packet
         E - detection packet 
         C - calibration packet 
         T - timing packet 
         L - log packet 
         O - opaque data packet 
         U - unknown/general packet 
         I - INFO packet
         ? - unidentifiable packet
.fi

The flags are prefaced with either the \fB%\fP or \fB#\fP modifier.
The \fB%\fP modifier indicates a defining flag while the \fB#\fP
indicates a non-defining flag.  All received packets with the same set
of defining flags will be saved to the same file. Non-defining flags
will be expanded using the values in the first packet received for the
resulting file name.

Time flags are based on the start time of the given packet.

For example, the format string:

\fB/archive/%n/%s/%n.%s.%l.%c.%Y.%j\fP

would be expanded to day length files named something like:

\fB/archive/NL/HGN/NL.HGN..BHE.2003.055\fP

Using non-defining flags the format string:

\fB/data/%n.%s.%Y.%j.%H:#M:#S.miniseed\fP

would be expanded to:

\fB/data/NL.HGN.2003.044.14:17:54.miniseed\fP

resulting in hour length files because the minute and second are
specified with the non-defining modifier.  The minute and second
fields are from the first packet in the file.

.SH "Stream list file"
The stream list file used with the '-l' option is expected to
define a data stream on each line.  The format of each line is:

.nf
<net> <station> [selectors]
.fi

The selectors are optional.  If default selectors are also specified
(with the '-s' option), they they will be used when no selectors are
specified for a given stream.  An example file follows:

.nf
----  Begin example file -----
# Comment lines begin with a '#' or '*'
# Example stream list file for use with the -l argument of slclient or
# with the sl_read_streamlist() libslink function.
GE ISP  BH?.D
NL HGN
MN AQU  BH? HH?
----  End example file -----
.fi

.SH "CAVEATS"

The future data checking options (-Fi and -Fc) only control the
writing of waveform data to archive files.  Any duplicates of other
packet types sent by the server will be written to their associated
archive file.

The future data checking options (-Fi and -Fc) are only consistent for
unique data streams written to a single archive file; in other words,
the checks do not span across different archive files.  As an example,
the SDS format creates "day files" which rotate at midnight.  The
future checks will not function correctly if there is a time jump
starting with the first packet in a new day file at midnight.  The
chance of this occurring is very, very low, but the behavior should be
noted nonetheless.

The initial future check (-Fi) triggers slarchive to perform a future
check whenever a file is opened.  An existing file can be re-opened if
it has been closed due to an idle stream timeout.

In general it is not a good idea to use non-defining modifier flags
(those starting with '#') in directory names.  Doing this with data
streams that are closed due to timeout and re-opened or slarchive
restarts will result in empty directories being created.  This is such
a fringe case that it will not be addressed any time soon.

.SH "SEE ALSO"
\fBslinktool\fP(1)

.SH "NOTES"
SeedLink is currently distributed as part of the SeisComP
(Seismological Communication Processor) package.  For more
information see, http://www.gfz-potsdam.de/geofon/

.SH AUTHOR
.nf
Chad Trabant
ORFEUS Data Center/EC-Project MEREDIAN
.fi


src/seedlink/apps/3rd-party/slarchive/src/slarchive.c
/***************************************************************************
 * slarchive.c
 * A SeedLink client for data collection and archiving.
 *
 * Connects to a SeedLink server, collects and archives data.
 *
 * Written by Chad Trabant, ORFEUS/EC-Project MEREDIAN
 *
 * modified 2005.147
 ***************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <time.h>

#include <libslink.h>

#include "archive.h"

#define PACKAGE   "slarchive"
#define VERSION   "1.7-sc3"

static void packet_handler (char *msrecord, int packet_type,
			    int seqnum, int packet_size);
static int  parameter_proc (int argcount, char **argvec);
static char *getoptval (int argcount, char **argvec, int argopt);
static int  add_dsarchive(const char *path, int archivetype);
static int  set_dsbuffersize(int size);
static void term_handler (int sig);
static void print_timelog (const char *msg);
static void usage (void);

/* A chain of archive definitions */
typedef struct DSArchive_s {
  DataStream  datastream;
  struct DSArchive_s *next;
}
DSArchive;

static short int verbose  = 0;   /* flag to control general verbosity */
static short int ppackets = 0;   /* flag to control printing of data packets */
static int stateint       = 0;   /* packet interval to save statefile */
static char *statefile    = 0;	 /* state file for saving/restoring stream states */

static SLCD *slconn;	         /* connection parameters */
static DSArchive *dsarchive;

int
main (int argc, char **argv)
{
  SLpacket *slpack;
  int seqnum;
  int ptype;
  int packetcnt = 0;

  /* Signal handling, use POSIX calls with standardized semantics */
  struct sigaction sa;

  sa.sa_flags   = SA_RESTART;
  sigemptyset (&sa.sa_mask);

  sa.sa_handler = term_handler;
  sigaction (SIGINT, &sa, NULL);
  sigaction (SIGQUIT, &sa, NULL);
  sigaction (SIGTERM, &sa, NULL);

  sa.sa_handler = SIG_IGN;
  sigaction (SIGHUP, &sa, NULL);
  sigaction (SIGPIPE, &sa, NULL);

  dsarchive = NULL;

  /* Allocate and initialize a new connection description */
  slconn = sl_newslcd();

  /* Process given parameters (command line and parameter file) */
  if (parameter_proc (argc, argv) < 0)
    {
      fprintf (stderr, "Parameter processing failed\n");
      fprintf (stderr, "Try '-h' for detailed help\n");
      return -1;
    }

  /* Loop with the connection manager */
  while ( sl_collect (slconn, &slpack) )
    {
      ptype  = sl_packettype (slpack);
      seqnum = sl_sequence (slpack);

      packet_handler ((char *) &slpack->msrecord, ptype, seqnum, SLRECSIZE);

      if ( statefile && stateint )
	{
	  if ( ++packetcnt >= stateint )
	    {
	      if (dsarchive) {
	        DSArchive *curdsa = dsarchive;

	        while ( curdsa != NULL ) {
		  if ( curdsa->datastream.grouphash ) {
                    DataStreamGroup *curgroup, *tmp;
                    HASH_ITER(hh, curdsa->datastream.grouphash, curgroup, tmp) {
		      if ( curgroup->filed ) {
			if ( curgroup->bp ) {
			  sl_log (1, 3, "Writing data to data stream file %s\n", curgroup->filename);
			  if ( write (curgroup->filed, curgroup->buf, curgroup->bp) != curgroup->bp ) {
			    sl_log (2, 1,
				    "main: failed to write record\n");
			    return -1;
			  }
			}
			curgroup->bp = 0;
		      }
		    }
		  }
	          curdsa = curdsa->next;
	        }
	      }
	      sl_savestate (slconn, statefile);
	      packetcnt = 0;
	    }
	}
    }

  /* Do all the necessary cleanup and exit */
  if (slconn->link != -1)
    sl_disconnect (slconn);
  
  if (dsarchive) {
    DSArchive *curdsa = dsarchive;

    while ( curdsa != NULL ) {
      archstream_proc (&curdsa->datastream, NULL, 0);
      curdsa = curdsa->next;
    }    
  }

  if (statefile)
    sl_savestate (slconn, statefile);

  return 0;
}  /* End of main() */


/***************************************************************************
 * packet_handler:
 * Process a received packet based on packet type.
 ***************************************************************************/
static void
packet_handler (char *msrecord, int packet_type, int seqnum, int packet_size)
{
  static SLMSrecord * msr = NULL;

  double dtime;			/* Epoch time */
  double secfrac;		/* Fractional part of epoch time */
  time_t ttime;			/* Integer part of epoch time */
  char   timestamp[20];
  struct tm *timep;
  int    archflag = 1;

  /* The following is dependent on the packet type values in libslink.h */
  char *type[]  = { "Data", "Detection", "Calibration", "Timing",
		    "Message", "General", "Request", "Info",
                    "Info (terminated)", "KeepAlive" };

  if ( verbose >= 1 ) {
    /* Build a current local time string */
    dtime   = sl_dtime ();
    secfrac = (double) ((double)dtime - (int)dtime);
    ttime   = (time_t) dtime;
    timep   = localtime (&ttime);
    snprintf (timestamp, 20, "%04d.%03d.%02d:%02d:%02d.%01.0f",
	      timep->tm_year + 1900, timep->tm_yday + 1, timep->tm_hour,
	      timep->tm_min, timep->tm_sec, secfrac);
    
    sl_log (1, 1, "%s, seq %d, Received %s blockette\n",
	    timestamp, seqnum, type[packet_type]);
  }

  /* Parse data record and print requested detail if any */
  if ( sl_msr_parse (slconn->log, msrecord, &msr, 1, 0) == NULL )
    {
      sl_log (2, 0, "%s, seq %d, Received %s blockette, error: blockette discarded\n",
                          timestamp, seqnum, type[packet_type]);
      return;
    }
  
  if ( ppackets )
    sl_msr_print (slconn->log, msr, ppackets - 1);
  
  /* Process waveform data and send it on */
  if ( packet_type == SLDATA )
    {
      /* Test for a so-called end-of-detection record */
      if ( msr->fsdh.samprate_fact == 0 && msr->fsdh.num_samples == 0 )
	archflag = 0;
    }
  
  /* Write packet to all archives in archive definition chain */
  if ( dsarchive && archflag ) {
    DSArchive *curdsa = dsarchive;

    while ( curdsa != NULL ) {
      curdsa->datastream.packettype = packet_type;

      /* Limit BUD archiving to waveform data */
      if ( curdsa->datastream.archivetype == BUD &&
	   packet_type != SLDATA ) {
	curdsa = curdsa->next;
	continue;
      }

      archstream_proc (&curdsa->datastream, msr, packet_size);
      
      curdsa = curdsa->next;
    }
  }
}  /* End of packet_handler() */


/***************************************************************************
 * parameter_proc:
 *
 * Process the command line parameters.
 *
 * Returns 0 on success, and -1 on failure.
 ***************************************************************************/
static int
parameter_proc (int argcount, char **argvec)
{
  int futurecontflag= 0;  /* continuous future check flag */
  int futurecont    = 2;  /* continuous future check overlap */
  int futureinitflag= 0;  /* initial future check (opening files) flag */
  int futureinit    = 2;  /* initial future check (opening files) overlap */
  int idletimeout   = 300; /* idle stream timeout */
  int error = 0;

  char *streamfile  = 0;   /* stream list file for configuring streams */
  char *multiselect = 0;
  char *selectors   = 0;
  char *timewin     = 0;
  char *tptr;

  SLstrlist *timelist;	   /* split the time window arg */

  if (argcount <= 1)
    error++;

  /* Process all command line arguments */
  for (optind = 1; optind < argcount; optind++)
    {
      if (strcmp (argvec[optind], "-V") == 0)
        {
          fprintf(stderr, "%s version: %s\n", PACKAGE, VERSION);
          exit (0);
        }
      else if (strcmp (argvec[optind], "-h") == 0)
        {
          usage();
          exit (0);
        }
      else if (strncmp (argvec[optind], "-v", 2) == 0)
	{
	  verbose += strspn (&argvec[optind][1], "v");
	}
      else if (strncmp (argvec[optind], "-p", 2) == 0)
        {
          ppackets += strspn (&argvec[optind][1], "p");
        }
      else if (strncmp (argvec[optind], "-Fc", 3) == 0)
        {
	  futurecontflag = 1;
	  if ( (tptr = strchr(argvec[optind], ':')) )
	    futurecont = atoi(tptr+1);
        }
      else if (strncmp (argvec[optind], "-Fi", 3) == 0)
        {
	  futureinitflag = 1;
	  if ( (tptr = strchr(argvec[optind], ':')) )
	    futureinit = atoi(tptr+1);
        }
      else if (strcmp (argvec[optind], "-d") == 0)
	{
	  slconn->dialup = 1;
	}
      else if (strcmp (argvec[optind], "-b") == 0)
	{
	  slconn->batchmode = 1;
	}
      else if (strcmp (argvec[optind], "-nt") == 0)
	{
	  slconn->netto = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-nd") == 0)
	{
	  slconn->netdly = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-k") == 0)
	{
	  slconn->keepalive = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-i") == 0)
	{
	  idletimeout = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-A") == 0)
	{
	  if ( add_dsarchive(getoptval(argcount, argvec, optind++), ARCH) == -1 )
	    return -1;
	}
	  else if (strcmp (argvec[optind], "-B") == 0)
	{
	  if ( set_dsbuffersize(atoi (getoptval(argcount, argvec, optind++))) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-SDS") == 0)
	{
	  if ( add_dsarchive(getoptval(argcount, argvec, optind++), SDS) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-BUD") == 0)
	{
	  if ( add_dsarchive(getoptval(argcount, argvec, optind++), BUD) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-DLOG") == 0)
	{
	  if ( add_dsarchive(getoptval(argcount, argvec, optind++), DLOG) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-l") == 0)
	{
	  streamfile = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-s") == 0)
	{
	  selectors = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-S") == 0)
	{
	  multiselect = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-x") == 0)
	{
	  statefile = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-tw") == 0)
	{
	  timewin = getoptval(argcount, argvec, optind++);
	}
      else if (strncmp (argvec[optind], "-", 1) == 0)
	{
	  fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
	  exit (1);
	}
      else if (!slconn->sladdr)
        {
          slconn->sladdr = argvec[optind];
        }
      else
	{
	  fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
	  exit (1);
	}
    }

  /* Make sure a server was specified */
  if ( ! slconn->sladdr )
    {
      fprintf(stderr, "No SeedLink server specified\n\n");
      fprintf(stderr, "%s version %s\n\n", PACKAGE, VERSION); 
      fprintf(stderr, "Usage: %s [options] [host][:][port]\n\n", PACKAGE);
      fprintf(stderr, "Try '-h' for detailed help\n");
      exit (1);
    }
  
  /* Initialize the verbosity for the sl_log function */
  sl_loginit (verbose, &print_timelog, NULL, &print_timelog, NULL);

  /* Set stdout (where logs go) to always flush after a newline */
  setvbuf(stdout, NULL, _IOLBF, 0);

  /* Report the program version */
  sl_log (1, 1, "%s version: %s\n", PACKAGE, VERSION);

  /* If errors then report the usage message and quit */
  if (error)
    {
      usage ();
      exit (1);
    }
  
  /* Load the stream list from a file if specified */
  if ( streamfile )
    sl_read_streamlist (slconn, streamfile, selectors);

  /* Split the time window argument */
  if ( timewin )
    {

      if (strchr (timewin, ':') == NULL)
	{
	  sl_log (2, 0, "time window not in begin:[end] format\n");
	  return -1;
	}

      if (sl_strparse (timewin, ":", &timelist) > 2)
	{
	  sl_log (2, 0, "time window not in begin:[end] format\n");
	  sl_strparse (NULL, NULL, &timelist);
	  return -1;
	}

      if (strlen (timelist->element) == 0)
	{
	  sl_log (2, 0, "time window must specify a begin time\n");
	  sl_strparse (NULL, NULL, &timelist);
	  return -1;
	}

      slconn->begin_time = strdup (timelist->element);

      timelist = timelist->next;

      if (timelist != 0)
	{
	  slconn->end_time = strdup (timelist->element);

	  if (timelist->next != 0)
	    {
	      sl_log (2, 0, "malformed time window specification\n");
	      sl_strparse (NULL, NULL, &timelist);
	      return -1;

	    }
	}

      /* Free the parsed list */
      sl_strparse (NULL, NULL, &timelist);
    }

  /* Parse the 'multiselect' string following '-S' */
  if ( multiselect )
    {
      if ( sl_parse_streamlist (slconn, multiselect, selectors) == -1 )
	return -1;
    }
  else if ( !streamfile )
    {		         /* No 'streams' array, assuming uni-station mode */
      sl_setuniparams (slconn, selectors, -1, 0);
    }

  /* Attempt to recover sequence numbers from state file */
  if (statefile)
    {
      /* Check if interval was specified for state saving */
      if ((tptr = strchr (statefile, ':')) != NULL)
	{
	  char *tail;
	  
	  *tptr++ = '\0';
	  
	  stateint = (unsigned int) strtoul (tptr, &tail, 0);
	  
	  if ( *tail || (stateint < 0 || stateint > 1e9) )
	    {
	      sl_log (2, 0, "state saving interval specified incorrectly\n");
	      return -1;
	    }
	}

      if (sl_recoverstate (slconn, statefile) < 0)
	{
	  sl_log (2, 0, "state recovery failed\n");
	}
    }
  
  /* If no archiving is specified print a warning */
  if ( !dsarchive ) {
    sl_log (1, 0, "WARNING: no archiving method was specified\n");
  }
  /* Otherwise fill in the global parameters for each entry */
  else {
    DSArchive *curdsa = dsarchive;
    while ( curdsa != NULL ) {
      curdsa->datastream.idletimeout = idletimeout;
      curdsa->datastream.futurecontflag = futurecontflag;
      curdsa->datastream.futurecont = futurecont;
      curdsa->datastream.futureinitflag = futureinitflag;
      curdsa->datastream.futureinit = futureinit;
      curdsa = curdsa->next;
    }
  }

  return 0;
}  /* End of parameter_proc() */


/***************************************************************************
 * getoptval:
 *
 * Return the value to a command line option; checking that the value is 
 * itself not an option (starting with '-') and is not past the end of
 * the argument list.
 *
 * argcount: total arguments in argvec
 * argvec: argument list
 * argopt: index of option to process, value is expected to be at argopt+1
 *
 * Returns value on success and exits with error message on failure
 ***************************************************************************/
static char *
getoptval (int argcount, char **argvec, int argopt)
{
  if ( argvec == NULL || argvec[argopt] == NULL ) {
    fprintf (stderr, "getoptval(): NULL option requested\n");
    exit (1);
  }

  if ( (argopt+1) < argcount && *argvec[argopt+1] != '-' )
    return argvec[argopt+1];

  fprintf (stderr, "Option %s requires a value\n", argvec[argopt]);
  exit (1);
}  /* End of getoptval() */


/***************************************************************************
 * add_dsarchive():
 * Add entry to the data stream archive chain.
 *
 * Returns 0 on success, and -1 on failure
 ***************************************************************************/
static int
add_dsarchive( const char *path, int archivetype )
{
  DSArchive *newdsa;

  newdsa = (DSArchive *) malloc (sizeof (DSArchive));
  
  if ( newdsa == NULL ) {
    sl_log (2, 0, "cannot allocate memory for new archive definition\n");
    return -1;
  }

  /* Setup new entry and add it to the front of the chain */
  newdsa->datastream.path = strdup(path);
  newdsa->datastream.archivetype = archivetype;
  newdsa->datastream.grouphash = NULL;

  if ( newdsa->datastream.path == NULL ) {
    sl_log (2, 0, "cannot allocate memory for new archive path\n");
    return -1;
  }

  newdsa->next = dsarchive;
  dsarchive = newdsa;

  return 0;
}  /* End of add_dsarchive() */


/***************************************************************************
 * set_dsbuffersize():
 * Sets the buffer size of the archive
 *
 * Returns 0 on success, and -1 on failure
 ***************************************************************************/
static int
set_dsbuffersize( int size )
{
  if ( size < 0 ) return -1;
  DS_BUFSIZE = size * 512;
  return 0;
}


/***************************************************************************
 * term_handler:
 * Signal handler routine to controll termination.
 ***************************************************************************/
static void
term_handler (int sig)
{
  sl_terminate (slconn);
}


/***************************************************************************
 * print_timelog:
 * Log message print handler used with sl_loginit() and sl_log().  Prefixes
 * a local time string to the message before printing.
 ***************************************************************************/
static void
print_timelog (const char *msg)
{
  char timestr[100];
  time_t loc_time;
  
  /* Build local time string and cut off the newline */
  time(&loc_time);
  strcpy(timestr, asctime(localtime(&loc_time)));
  timestr[strlen(timestr) - 1] = '\0';
  
  fprintf (stdout, "%s - %s", timestr, msg);
}


/***************************************************************************
 * usage:
 * Print the usage message and exit.
 ***************************************************************************/
static void
usage (void)
{
  fprintf (stderr, "%s version %s\n\n", PACKAGE, VERSION);
  fprintf (stderr, "Usage: %s [options] [host][:][port]\n\n", PACKAGE);
  fprintf (stderr,
	   " ## General program options ##\n"
	   " -V              report program version\n"
	   " -h              show this usage message\n"
	   " -v              be more verbose, multiple flags can be used\n"
	   " -p              print details of data packets, multiple flags can be used\n"
	   " -nd delay       network re-connect delay (seconds), default 30\n"
	   " -nt timeout     network timeout (seconds), re-establish connection if no\n"
	   "                   data/keepalives are received in this time, default 600\n"
	   " -k interval     send keepalive (heartbeat) packets this often (seconds)\n"
	   " -x sfile[:int]  save/restore stream state information to this file\n"
	   " -i timeout      idle stream entries might be closed (seconds), default 300\n"
	   " -d             configure the connection in dial-up mode\n"
	   " -b             configure the connection in batch mode\n"
	   " -Fi[:overlap]   Initially check (existing files) that data records are newer\n"
	   " -Fc[:overlap]   Continuously check that data records are newer\n"
	   "\n"
	   " ## Data stream selection ##\n"
	   " -s selectors    selectors for uni-station or default for multi-station mode\n"
	   " -l listfile     read a stream list from this file for multi-station mode\n"
           " -S streams      define a stream list for multi-station mode\n"
	   "   'streams' = 'stream1[:selectors1],stream2[:selectors2],...'\n"
	   "        'stream' is in NET_STA format, for example:\n"
	   "        -S \"IU_KONO:BHE BHN,GE_WLF,MN_AQU:HH?.D\"\n\n"
	   " -tw begin:[end]  (requires SeedLink >= 3)\n"
	   "        specify a time window in year,month,day,hour,min,sec format\n"
	   "        example: -tw 2002,08,05,14,00,00:2002,08,05,14,15,00\n"
	   "        the end time is optional, but the colon must be present\n"
	   "\n"
	   " ## Data archiving options ##\n"
	   " -A format       save all received records is a custom file structure\n"
	   " -B count        set record buffer size to count*512, default 1000\n"
	   " -SDS  SDSdir    save all received records in a SDS file structure\n"
	   " -BUD  BUDdir    save all received data records in a BUD file structure\n"
	   " -DLOG DLOGdir   save all received data records in an old-style\n"
	   "                   SeisComP/datalog file structure\n"
	   "\n"
	   " [host][:][port] Address of the SeedLink server in host:port format\n"
	   "                  Default host is 'localhost' and default port is '18000'\n\n");

}  /* End of usage() */
